### DISCUSSION TITLE
getting the error       CS1929	'WebApplication' does not contain a definition for 'UseWorkflows' and the best extension method overload 'ModuleExtensions.UseWorkflows(IModule, Action<WorkflowsFeature>?)' requires a receiver of type 'Elsa.Features.Services.IModule'

### DISCUSSION THREAD

**Discussion by Chandrasanni** [elsa-workflows/elsa-core#5299] [Q&A]
Hi Team,

I have tried using the https://github.com/elsa-workflows/elsa-core/tree/main/samples/aspnet/Elsa.Samples.AspNet.QuartzIntegration to implement in my project. after installing the latest version of Elsa 3.1.3. I facing the below error message.

CS1929	'WebApplication' does not contain a definition for 'UseWorkflows' and the best extension method overload 'ModuleExtensions.UseWorkflows(IModule, Action<WorkflowsFeature>?)' requires a receiver of type 'Elsa.Features.Services.IModule'

at Program.cs file as below
var builder = WebApplication.CreateBuilder(args);

var environment = Environment.GetEnvironmentVariable("ASPNETCORE_ENVIRONMENT") ?? "Development";
var configuration = new ConfigurationBuilder()
    .SetBasePath(Directory.GetCurrentDirectory())
    .AddJsonFile($"appsettings.{environment}.json", optional: true, reloadOnChange: true)
    .AddEnvironmentVariables()
    .Build();

// Add services to the container.
builder.Host.UseSerilog((context, loggerConfig) =>
{
// some detials;
});
var myAllowSpecificOrigins = Constants.AllowSpecificOriginsKey;

var app = builder.Build();

if (app.Environment.IsDevelopment())
{
    app.UseSwagger();
    app.UseSwaggerUI();
}

app.UseHttpsRedirection();
app.UseCors(myAllowSpecificOrigins);

app.UseExceptionHandler("/error");
app.UseHsts();
app.UseExceptionHandler();

app.UseAuthorization();
app.UseWorkflowsApi();
app.UseWorkflows();

app.UseSerilogRequestLogging(options =>
{
    options.EnrichDiagnosticContext = Enricher.HttpRequestEnricher;
});
app.MapControllers();

app.Run();

**Comment by glime-ai** [comment]
Based on the context provided and the error message you're encountering, it seems like there might be a mismatch between the version of Elsa you are using and the method calls in your `Program.cs` file. The `UseWorkflows` extension method might not be available or might have been changed in Elsa 3.1.3.

In Elsa 3.x, the way to set up the application has changed compared to previous versions. Instead of using `UseWorkflows`, you should configure your application using the `AddElsa` extension method on the `IServiceCollection` provided by the `builder.Services`. This method is used to add and configure services required by Elsa.

Here's an example of how you might set up your `Program.cs` based on the context provided:

```csharp
var builder = WebApplication.CreateBuilder(args);

// Configure services here...
builder.Services.AddElsa(elsa => {
    // Configure Elsa services as needed, for example:
    elsa.AddHttpActivities();
    // ... other configurations
});

var app = builder.Build();

// Configure the HTTP request pipeline here...
app.UseHttpsRedirection();
app.UseCors();

app.UseAuthorization();
app.UseElsaEndpoints(); // This replaces app.UseWorkflowsApi() and app.UseWorkflows()

// ... other middleware configurations

app.Run();
```

Please note that `UseElsaEndpoints` is a method that sets up the necessary endpoints for Elsa, including the API and workflow endpoints. This method should replace the `app.UseWorkflowsApi()` and `app.UseWorkflows()` calls.

Make sure to check the official Elsa documentation for the version you are using to get the correct setup instructions. If the documentation for version 3.1.3 is not available, you may need to look at the source code or examples from the same version to understand the correct way to configure your application.

If you continue to face issues, consider reaching out to the Elsa community or checking the GitHub repository for issues related to the version you are using.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=dRbvTb7IhCSYiI&utm_medium=github).

**Reply by Chandrasanni** [reply]
But, I did not observe the UseElsaEndPoints() and AddHttpActivities()

**Reply by Chandrasanni** [reply]
Is the 3.1 having those UseElsaEndPoints() and AddHttpActivities()?

**Reply by Chandrasanni** [reply]
This is not working in my case. I have modified the code
builder.Services.AddElsa(elsa => {
    elsa.UseWorkflowManagement(management => management.UseEntityFrameworkCore());
    elsa.UseWorkflowRuntime(runtime =>
    {
        runtime.UseEntityFrameworkCore();
    });
    // Expose API endpoints.
    elsa.UseQuartz();
    elsa.UseScheduling(scheduling => scheduling.UseQuartzScheduler());
    elsa.AddWorkflowsFrom<Program>();
    elsa.AddHttpActivities();
});
AddHttpActivities is not CS1061	'IModule' does not contain a definition for 'AddHttpActivities' and no accessible extension method 'AddHttpActivities' accepting a first argument of type 'IModule' could be found (are you missing a using directive or an assembly reference?)

And app.UseElsaEndPoints();
Error CS1061	'WebApplication' does not contain a definition for 'UseElsaEndPoints' and no accessible extension method 'UseElsaEndPoints' accepting a first argument of type 'WebApplication' could be found (are you missing a using directive or an assembly reference?)
